### ISSUE TITLE
Screenshot not taken in time before next action

### ISSUE THREAD

**Issue by bhoppeadoy** [azachar/protractor-screenshoter-plugin#73]
Bug report
---
- Plugin Version: `0.10.1`
- Protractor Version: `5.3.2`
- Browser(s): `Chrome 67`
- Node Version: `8.9.4`
- Angular Version: `5.2.0`
- Operating System and Version `RHEL 7.4`

```
 plugins: [{
    package: 'protractor-screenshoter-plugin',
    screenshotPath: './test-results/e2e/html-report',
    screenshotOnExpect: 'failure',
    screenshotOnSpec: 'none',
    htmlOnExpect: 'failure',
    htmlOnSpec: 'none',
    withLogs: false,
    writeReportFreq: 'asap',
    imageToAscii: 'none',
    clearFoldersBeforeTest: true
  }],
```
---

I've been using the protractor-screenshoter-plugin and its been great so far. 95% of the time it works fine. Every once and a while I get a scenario where the screenshot taken for a failed expect is not of the screen where the failure happened, but of the screen after the next action has already been done.

For example, this dummy scenario:

```
await browser.get('/page2');
expect(isPage1).toBe(true, 'Not page1');
await browser.get('/page3');
```

1) The test goes to page2. 
2) Do an expect() to check if its page1. This obviously fails and should take a screen shot where it would show page2 in the image. 
3) What happens is the screenshot is taken after its gone to page3.

Now, I believe this may be an issue with await/async vs control flow. I don't recall seeing this when I was using control flow in the test. Now that I converted my test to use await/async its happened several times.

Perhaps since the reporter attaches to the expect() call, that the test doesn't wait on the reporter to finish before moving on to the next line.

**Comment by azachar** [comment]
Hi there,
 thank you for your detailed report I really appreciate it.


 I’ve seen in your configuration file you are using writeReportFreq: ‘asap’. Try to use the default value which is ‘end’, it could save  some processing time.


 If this doesn’t help you might be completely right with your suggestion.  Could you please create a test that is failing?  I already have some tests inside the plugin, just copy& paste one and adjust to your scenario.

Cheers,
Andrej

**Comment by bhoppeadoy** [comment]
@azachar 

Thanks. I actually have it set with writeReportFreq: 'spec', but I'll try it with 'end' to see if it helps

Reproducing the issue is difficult but I'll see if I can create a test case if the above doesn't work.

**Comment by azachar** [comment]
Sure, let me know how it went.

Well, the plugin needs to be rewritten for fully promised based intercepting, there are still some hacks that are smelly to me. And your case might be the one of them.

**Comment by azachar** [comment]
Now I just released a new version, let mw know if the issues still persists. For time being I am closing this issue.

**Comment by coyoteecd** [comment]
@azachar 
Please reopen this, I am able to reproduce it using the latest 0.10.3 version and async/await as @bhoppeadoy described.

A similar piece of dummy code:
```
expect(await browser.getCurrentUrl()).toBeDefined();
await browser.close();
```
... will give errors when taking screenshots/logs/HTML data because the browser window is already closed.

The cause is likely the implementation of forEachBrowser that first schedules a command to get the capabilities in order to read the browser name. This command is a promise that is never awaited/returned, therefore the actual taking of the screenshot will execute "some time later, after the capabilities are retrieved".

And because the expect is synchronously executed, as soon as that promise is scheduled, jasmine will move on to schedule the next command which is the `browser.close()`. The end result of webdriver commands that get sent is:
- get current url
- get capabilities
- close
- take screenshot

For me the quick fix was to simply remove the getCapabilities part from forEachBrowser and use a hard-coded name for the global browser:
```
protractorUtil.forEachBrowser = function(action) {
  function catchError(err) {
    console.warn('Unknown error:');
    console.warn(err);
  }

  try {
    if (global.screenshotBrowsers && Object.keys(global.screenshotBrowsers).length > 0) {
      _.forOwn(global.screenshotBrowsers, function(instance, name) {
        action(instance, name, protractorUtil.newLongRunningOperationCounter());
      });
    } else {
      action(global.browser, 'mainBrowser', protractorUtil.newLongRunningOperationCounter());
    }
  } catch (err) {
    catchError(err);
  }
};
```

An alternate solution could be to take the browser name *after* you've taken the screenshot/log/page source.

**Comment by azachar** [comment]
Great catch, thank you!
Sure, this needs to be fixed.

**Comment by greimela** [comment]
This issue is still reproducible in the current version.
The hack from @coyoteecd works but it would be nice to have this fixed upstream!
